typings/styled-components.d.ts

```diff
@@ -24,9 +24,100 @@ export interface InterpolationFunction<P> {
   (props: P): Interpolation<P>;
 }
 
+export interface StyledComponentClass<P, T> extends ComponentClass<P> {
+  extend: ThemedStyledFunction<P, T>;
+
+  withComponent(
```

Comment:
Regarding [your question](https://github.com/styled-components/styled-components/pull/837#issuecomment-304453938), I think we can very efficiently use mapped types here.
Say, we define all the tags and corresponding props as following:
```ts
interface HtmlTags {
  a: HTMLAnchorElement;
  area: HTMLAreaElement;
  // ..., svg
}
```
then define like this:
```ts
type WithComponentOverloads<T> = {
  [K in keyof HtmlTags]: (tag: K) => StyledComponentClass<HtmlTags[K], T>;
};
interface StyledComponent<T> {
  withComponent: WithComponentOverloads<T>[keyof HtmlTags];
  //...
}
```
same `HtmlTags` can also be used in declaration of `ThemedStyledInterface`.

I'm not really sure how well it's gonna work. I can try to experiment with that.